### ISSUE TITLE
docker_image async task fails with 104 Connection reset by peer

### ISSUE THREAD

**Issue by pkaramol** [ansible/ansible#47065]
##### SUMMARY
Waiting on an async task fails with requests `104` error

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
core 

##### ANSIBLE VERSION

```paste below
ansible 2.7.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/pkara/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]

```

##### CONFIGURATION

```paste below
*$ ansible-config dump --only-changed
DEFAULT_ROLES_PATH(/home/pkara/Workspace/gitlab/mdr-deployment/ansible/ansible.cfg) = [u'/home/pkara/Workspace/gitlab/mdr-deployment/ansible/roles']
DEFAULT_STDOUT_CALLBACK(/home/pkara/Workspace/gitlab/mdr-deployment/ansible/ansible.cfg) = yaml
DEFAULT_VAULT_PASSWORD_FILE(/home/pkara/Workspace/gitlab/mdr-deployment/ansible/ansible.cfg) = /home/pkara/Workspace/gitlab/mdr-deployment/ansible/.vault
RETRY_FILES_ENABLED(/home/pkara/Workspace/gitlab/mdr-deployment/ansible/ansible.cfg) = False

```

##### OS / ENVIRONMENT

```
$ cat /etc/os-release 
NAME="Ubuntu"
VERSION="16.04.5 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.5 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
/home/pkara/Workspace/gitlab/mdr-deployment/ansible
$ uname -a
Linux pkara-pc01 4.15.0-36-generic #39~16.04.1-Ubuntu SMP Tue Sep 25 08:59:23 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

##### STEPS TO REPRODUCE

Here is a task called in `async` mode

```yaml
 - name: Build Images --> Build DB image
    docker_image:
      path: "{{ temp_build_directory }}/api/db"
      name: "{{ docker_registry }}/db"
      tag: "{{ docker_tag }}"
      state: present
      force: yes
      push: yes
    async: 10000
    poll: 0
    register: db_build
```

Here its' await counterpart:

```yaml
  - name: Await Image Build --> Await docker db image build
    async_status:
      jid: "{{ db_build.ansible_job_id }}"
    register: db_result
    until: db_result.finished
    retries: 150
    tags:
      - api
```

Here is the failure:

```python
TASK [build_and_retrieve_images : Await Image Build --> Await docker db image build] ***
task path: /home/pkara/Workspace/gitlab/mdr-deployment/ansible/roles/build_and_retrieve_images/tasks/await_image_build.yml:3
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: pkara
<localhost> EXEC /bin/sh -c 'echo ~pkara && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /var/tmp/ansible-tmp-1539617611.77-78021851949764 `" && echo ansible-tmp-1539617611.77-78021851949764="` echo /var/tmp/ansible-tmp-1539617611.77-78021851949764 `" ) && sleep 0'
Using module file /usr/lib/python2.7/dist-packages/ansible/modules/utilities/logic/async_status.py
<localhost> PUT /home/pkara/.ansible/tmp/ansible-local-17253v_brvl/tmpXmTb7x TO /var/tmp/ansible-tmp-1539617611.77-78021851949764/AnsiballZ_async_status.py
<localhost> EXEC /bin/sh -c 'setfacl -m u:ansible:r-x /var/tmp/ansible-tmp-1539617611.77-78021851949764/ /var/tmp/ansible-tmp-1539617611.77-78021851949764/AnsiballZ_async_status.py && sleep 0'
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u ansible /bin/sh -c '"'"'echo BECOME-SUCCESS-tpmkgfyumwirjapvkihnsqbrpibritnp; /usr/bin/python /var/tmp/ansible-tmp-1539617611.77-78021851949764/AnsiballZ_async_status.py'"'"' && sleep 0'
<localhost> EXEC /bin/sh -c 'rm -f -r /var/tmp/ansible-tmp-1539617611.77-78021851949764/ > /dev/null 2>&1 && sleep 0'
fatal: [vagrant -> localhost]: FAILED! => changed=false 
  ansible_job_id: '534008237800.18884'
  attempts: 1
  cmd: /var/tmp/ansible-tmp-1539617609.41-6376619420111/AnsiballZ_docker_image.py
  data: ''
  finished: 1
  invocation:
    module_args:
      jid: '534008237800.18884'
      mode: status
  msg: |-
    Traceback (most recent call last):
      File "/var/tmp/ansible-tmp-1539617609.41-6376619420111/async_wrapper.py", line 147, in _run_module
        (filtered_outdata, json_warnings) = _filter_non_json_lines(outdata)
      File "/var/tmp/ansible-tmp-1539617609.41-6376619420111/async_wrapper.py", line 88, in _filter_non_json_lines
        raise ValueError('No start of json char found')
    ValueError: No start of json char found
  stderr: |-
    Traceback (most recent call last):
      File "/var/tmp/ansible-tmp-1539617609.41-6376619420111/AnsiballZ_docker_image.py", line 113, in <module>
        _ansiballz_main()
      File "/var/tmp/ansible-tmp-1539617609.41-6376619420111/AnsiballZ_docker_image.py", line 105, in _ansiballz_main
        invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
      File "/var/tmp/ansible-tmp-1539617609.41-6376619420111/AnsiballZ_docker_image.py", line 48, in invoke_module
        imp.load_module('__main__', mod, module, MOD_DESC)
      File "/tmp/ansible_docker_image_payload_ZJgq0J/__main__.py", line 613, in <module>
      File "/tmp/ansible_docker_image_payload_ZJgq0J/__main__.py", line 608, in main
      File "/tmp/ansible_docker_image_payload_ZJgq0J/__main__.py", line 301, in __init__
      File "/tmp/ansible_docker_image_payload_ZJgq0J/__main__.py", line 329, in present
      File "/tmp/ansible_docker_image_payload_ZJgq0J/__main__.py", line 531, in build_image
      File "/usr/local/lib/python2.7/dist-packages/docker/api/build.py", line 259, in build
        timeout=timeout,
      File "/usr/local/lib/python2.7/dist-packages/docker/utils/decorators.py", line 46, in inner
        return f(self, *args, **kwargs)
      File "/usr/local/lib/python2.7/dist-packages/docker/api/client.py", line 194, in _post
        return self.post(url, **self._set_request_timeout(kwargs))
      File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 565, in post
        return self.request('POST', url, data=data, json=json, **kwargs)
      File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 518, in request
        resp = self.send(prep, **send_kwargs)
      File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 639, in send
        r = adapter.send(request, **kwargs)
      File "/usr/local/lib/python2.7/dist-packages/requests/adapters.py", line 488, in send
        raise ConnectionError(err, request=request)
    requests.exceptions.ConnectionError: ('Connection aborted.', error(104, 'Connection reset by peer'))
  stderr_lines:
  - 'Traceback (most recent call last):'
  - '  File "/var/tmp/ansible-tmp-1539617609.41-6376619420111/AnsiballZ_docker_image.py", line 113, in <module>'
  - '    _ansiballz_main()'
  - '  File "/var/tmp/ansible-tmp-1539617609.41-6376619420111/AnsiballZ_docker_image.py", line 105, in _ansiballz_main'
  - '    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)'
  - '  File "/var/tmp/ansible-tmp-1539617609.41-6376619420111/AnsiballZ_docker_image.py", line 48, in invoke_module'
  - '    imp.load_module(''__main__'', mod, module, MOD_DESC)'
  - '  File "/tmp/ansible_docker_image_payload_ZJgq0J/__main__.py", line 613, in <module>'
  - '  File "/tmp/ansible_docker_image_payload_ZJgq0J/__main__.py", line 608, in main'
  - '  File "/tmp/ansible_docker_image_payload_ZJgq0J/__main__.py", line 301, in __init__'
  - '  File "/tmp/ansible_docker_image_payload_ZJgq0J/__main__.py", line 329, in present'
  - '  File "/tmp/ansible_docker_image_payload_ZJgq0J/__main__.py", line 531, in build_image'
  - '  File "/usr/local/lib/python2.7/dist-packages/docker/api/build.py", line 259, in build'
  - '    timeout=timeout,'
  - '  File "/usr/local/lib/python2.7/dist-packages/docker/utils/decorators.py", line 46, in inner'
  - '    return f(self, *args, **kwargs)'
  - '  File "/usr/local/lib/python2.7/dist-packages/docker/api/client.py", line 194, in _post'
  - '    return self.post(url, **self._set_request_timeout(kwargs))'
  - '  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 565, in post'
  - '    return self.request(''POST'', url, data=data, json=json, **kwargs)'
  - '  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 518, in request'
  - '    resp = self.send(prep, **send_kwargs)'
  - '  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 639, in send'
  - '    r = adapter.send(request, **kwargs)'
  - '  File "/usr/local/lib/python2.7/dist-packages/requests/adapters.py", line 488, in send'
  - '    raise ConnectionError(err, request=request)'
  - 'requests.exceptions.ConnectionError: (''Connection aborted.'', error(104, ''Connection reset by peer''))'
```

##### EXPECTED RESULTS

Await task should keep retrying until the task is finished

##### ACTUAL RESULTS

Await task fails with the above error

**Comment by ansibot** [comment]
Hi @pkaramol, thank you for submitting this issue!

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by sivel** [comment]
This appears to be caused by the `docker_image` module causing a Traceback, which is not recoverable.

Pinging some docker contributors:

cc @dariko @felixfontein @jwitko @kassiansun @tbouvet

**Comment by felixfontein** [comment]
The error happens in `docker_image` as it tries to build the container. The code calls [client.build()](https://docker-py.readthedocs.io/en/stable/api.html#module-docker.api.build) and processes its output; the stacktrace is an internal error appearing inside docker-py (it is not one of the documented errors of the build function). It could be a timeout, or something else.

Searching for the underlying error (`requests.exceptions.ConnectionError: ('Connection aborted.', error(104, 'Connection reset by peer')`) on the net doesn't yield usable results. There are claims about similar errors ranging from "Python used doesn't support TLS 1.2, but server requires it" to "URL used is wrong".

Does `docker_image` works in other situations? Or any other of the `docker_*` modules (like `docker_image_facts` and `docker_container`)? If they work, this is probably a timeout. If not, it's probably a problem with connecting to the docker daemon.

!component =lib/ansible/modules/cloud/docker/docker_image.py

**Comment by felixfontein** [comment]
needs_info

**Comment by Akasurde** [comment]
!component =lib/ansible/modules/cloud/docker/docker_image.py

**Comment by ansibot** [comment]
@pkaramol This issue is waiting for your response. Please respond or the issue will be closed.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@pkaramol You have not responded to information requests in this issue so we will assume it no longer affects you. If you are still interested in this, please create a new issue with the requested information.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)
